### PR TITLE
Add some code examples to explain the usage of the format function in the String .

### PR DESCRIPTION
**Pull request by Rindbee**
Add some code examples to the format function of the String.

### CODE REVIEW

File: doc/classes/String.xml [hunk 1]
```diff
@@ -181,6 +181,25 @@
 			<argument index="1" name="placeholder" type="String" default="&quot;{_}&quot;" />
 			<description>
 				Formats the string by replacing all occurrences of [code]placeholder[/code] with [code]values[/code].
+				[b]Note:[/b] There must be one and only one [code]_[/code] in the [code]placeholder[/code].
+				[codeblocks]
+				[gdscript]
+				# Use the defalut placeholder, and a two-dimensional array.
+				var str1 := "name will be replaced to {name}, 0 maps {0}, strength is {strength}"
+				# Then the result1 is "name will be replaced to RobotGuy, 0 maps godot, strength is 9000.91"
```
Comment (YeldhamDev) [inline doc/classes/String.xml:189]: Needs an end dot here.

File: doc/classes/String.xml [hunk 2]
```diff
@@ -181,6 +181,25 @@
...
+				var result1 := origin.format([["name","RobotGuy"],[0,"godot"],["strength",9000.91]])
+
+				# Use a custom placeholder string $_, and a dictionary.
+				var str2 := "red=$red green=$green blue=$blue alpha=$alpha"
+				# Then the result2 is "red=0.1 green=0.3 blue=0.5 dict={hello:godoter}"
+				var result2 := origin.format({"red" : 0.1, "green" : 0.3, "blue" : 0.5, "dict" : {"hello" : "godoter"}}, "$_")
+
+				# Use an one-dimensional array. The value will be mapped by the index.
+				var str3 := "{1} is second，{0} is first, {2} is third"
+				# Will be "second is second，first is first, third is third"
```
Comment (YeldhamDev) [inline doc/classes/String.xml:199]: An end dot here too.

File: doc/classes/String.xml [hunk 3]
```diff
@@ -181,6 +181,25 @@
 			<argument index="1" name="placeholder" type="String" default="&quot;{_}&quot;" />
 			<description>
 				Formats the string by replacing all occurrences of [code]placeholder[/code] with [code]values[/code].
+				[b]Note:[/b] There must be one and only one [code]_[/code] in the [code]placeholder[/code].
+				[codeblocks]
+				[gdscript]
+				# Use the defalut placeholder, and a two-dimensional array.
+				var str1 := "name will be replaced to {name}, 0 maps {0}, strength is {strength}"
```
Comment (YeldhamDev) [inline doc/classes/String.xml:188]: Would be better if type inferring was not present here, as it isn't present in other examples across the docs.

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

File: doc/classes/String.xml (hunk 1 above)
Comment (Rindbee) [reply doc/classes/String.xml]: Modifications have been made according to your suggestion.

**Comment by YeldhamDev** [comment]
Good. Now, please [squash your commits](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase).

File: doc/classes/String.xml [hunk 4]
```diff
@@ -181,6 +181,25 @@
...
+				[gdscript]
+				# Use the defalut placeholder, and a two-dimensional array.
+				var str1 = "name will be replaced to {name}, 0 maps {0}, strength is {strength}"
+				# Then the result1 is "name will be replaced to RobotGuy, 0 maps godot, strength is 9000.91".
+				var result1 = str1.format([["name", "RobotGuy"], [0, "godot"], ["strength", 9000.91]])
+
+				# Use a custom placeholder string "$_", and a dictionary.
+				var str2 = "red=$red green=$green blue=$blue alpha=$alpha"
+				# Then the result2 is "red=0.1 green=0.3 blue=0.5 dict={hello:godoter}".
+				var result2 = str2.format({"red" : 0.1, "green" : 0.3, "blue" : 0.5, "dict" : {"hello" : "godoter"}}, "$_")
```
Comment (YeldhamDev) [inline doc/classes/String.xml:195]: There shouldn't be a space to the left of ":".

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

File: doc/classes/String.xml (hunk 4 above)
Comment (Rindbee) [reply doc/classes/String.xml:195]: It seems to work fine, but following your advice anyway.

Comment (YeldhamDev) [reply doc/classes/String.xml]: It's just codestyle thing. :wink: Also I meant just the spaces to the left, the right ones were fine.

File: doc/classes/String.xml [hunk 5]
```diff
@@ -181,6 +181,25 @@
 			<argument index="1" name="placeholder" type="String" default="&quot;{_}&quot;" />
 			<description>
 				Formats the string by replacing all occurrences of [code]placeholder[/code] with [code]values[/code].
+				[b]Note:[/b] There must be one and only one [code]_[/code] in the [code]placeholder[/code].
+				[codeblocks]
+				[gdscript]
+				# Use the defalut placeholder, and a two-dimensional array.
```
Comment (YeldhamDev) [inline doc/classes/String.xml:187]: Typo on "defalut".

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

File: doc/classes/String.xml (hunk 5 above)
Comment (Rindbee) [reply doc/classes/String.xml:187]: Reminds me of the time when I was learning English, well, not yet graduated in languages(including the native language) studies.

File: doc/classes/String.xml [hunk 6]
```diff
@@ -181,6 +181,25 @@
...
+				[gdscript]
+				# Use the default placeholder, and a two-dimensional array.
+				var str1 = "name will be replaced to {name}, 0 maps {0}, strength is {strength}"
+				# Then the result1 is "name will be replaced to RobotGuy, 0 maps godot, strength is 9000.91".
+				var result1 = str1.format([["name", "RobotGuy"], [0, "godot"], ["strength", 9000.91]])
+
+				# Use a custom placeholder string "$_", and a dictionary.
+				var str2 = "red=$red green=$green blue=$blue alpha=$alpha"
+				# Then the result2 is "red=0.1 green=0.3 blue=0.5 dict={hello:godoter}".
+				var result2 = str2.format({"red":0.1, "green":0.3, "blue":0.5, "dict":{"hello":"godoter"}}, "$_")
```
Comment (YeldhamDev) [inline doc/classes/String.xml:195]: Almost there, just place back the spaces to the **right** of the ":" characters here.

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

**Review by YeldhamDev — APPROVED** [review]

File: doc/classes/String.xml [hunk 7]
```diff
@@ -181,6 +181,25 @@
...
+				[gdscript]
+				# Use the default placeholder, and a two-dimensional array.
+				var str1 = "name will be replaced to {name}, 0 maps {0}, strength is {strength}"
+				# Then the result1 is "name will be replaced to RobotGuy, 0 maps godot, strength is 9000.91".
+				var result1 = str1.format([["name", "RobotGuy"], [0, "godot"], ["strength", 9000.91]])
+
+				# Use a custom placeholder string "$_", and a dictionary.
+				var str2 = "red=$red green=$green blue=$blue alpha=$alpha"
+				# Then the result2 is "red=0.1 green=0.3 blue=0.5 dict={hello:godoter}".
+				var result2 = str2.format({"red": 0.1, "green": 0.3, "blue": 0.5, "dict": {"hello": "godoter"}}, "$_")
```
Comment (YeldhamDev) [inline doc/classes/String.xml:195]: Wait, this appears to be incorrect. The actual result is `red=0.1 green=0.3 blue=0.5 alpha=$alpha`

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

File: doc/classes/String.xml (hunk 7 above)
Comment (YeldhamDev) [reply doc/classes/String.xml]: Alright, but what's the point of the dictionary now? :stuck_out_tongue:

Comment (Rindbee) [reply doc/classes/String.xml]: > Wait, this appears to be incorrect. The result from this is `red=0.1 green=0.3 blue=0.5 alpha=$alpha`.

Yeah, it is, when transcribing, I want to show the possible situation, that is, there is no need for an exact match.

Comment (Rindbee) [reply doc/classes/String.xml]: > Alright, but what's the point of the dictionary now? stuck_out_tongue

Matching may not be done all at once. There may be some more complicated cases here, but this is just a demonstration of the usage of this method `format`, not the `Dictionary`.

Comment (YeldhamDev) [reply doc/classes/String.xml]: I still don't quite get it. Is the point of the example to show it **not** formatting "$alpha"? Because if so, a simpler example than a dictionary would be better.

Comment (Rindbee) [reply doc/classes/String.xml]: > I still don't quite get it. Is the point of the example to show it **not** formatting "$alpha"? Because if so, a simpler example than a dictionary would be better.

For a better example : 

`... company id is {company_id}, named {company_name}, ...`

If you just want to get a brief company information, and already have a dictionary like this:

`company_info =  {"company_id": 001, "company_name": "A", "departments":{"name":"a", "director":"Tom", "members":{...}}}`

You can just use `format` without having to extract the dictionary.

Comment (YeldhamDev) [reply doc/classes/String.xml]: I see that. I just don't understand the point of the `"dict": {"hello": "godoter"}` dictionary.

Comment (Rindbee) [reply doc/classes/String.xml]: > I see that. I just don't understand the point of the `"dict": {"hello": "godoter"}` dictionary.

It's just that I didn't know what to write at that time, I just wrote it casually.

Just want to tell the reader that nested dictionaries can be used in the function.

Comment (Rindbee) [reply doc/classes/String.xml]: > there is no need for an exact match.

There is no need for an exact match.

Comment (YeldhamDev) [reply doc/classes/String.xml]: > Just want to tell the reader that nested dictionaries can be used in the function. There is no need for an exact match.

Then maybe it could be explicitly written in the comment that the dictionary doesn't need to match it exactly. It would both explain it better and give a purpose to that sub-dictionary.

File: doc/classes/String.xml [hunk 8]
```diff
@@ -181,6 +181,29 @@
 			<argument index="1" name="placeholder" type="String" default="&quot;{_}&quot;" />
 			<description>
 				Formats the string by replacing all occurrences of [code]placeholder[/code] with [code]values[/code].
+				[b]Note:[/b] There must be one and only one [code]_[/code] in the [code]placeholder[/code].
+				[b]Note:[/b] The [code]values[/code] can be a dictionary, an one-dimensional array, or a two-dimensional array. When it's a two-dimensional array, the length of the inner arrays must be [code]2[/code], or an error will be generated.
+				[b]Note:[/b] The function will iterate over the [code]values[/code]. For a dictionary, first, it will  replace the [code]_[/code] with the key; Then use the result to find the first matched word; At last, use the value to replace the result (see [method replace]); For an one-dimensional array, like the dictionary, but the index number acts as the key, the element of the array acts as the value; For a two-dimensional array, the first element of the inner array acts as the key, the second one acts as the value.
```
Comment (YeldhamDev) [inline doc/classes/String.xml:186]:
```suggestion
				[b]Note:[/b] The function will iterate over the [code]values[/code]. For a dictionary, first, it will replace the [code]_[/code] with the key; Then use the result to find the first matched word; At last, use the value to replace the result (see [method replace]); For an one-dimensional array, like the dictionary, but the index number acts as the key, the element of the array acts as the value; For a two-dimensional array, the first element of the inner array acts as the key, the second one acts as the value.
```

**Review by YeldhamDev — CHANGES_REQUESTED** [review]

**Comment by YeldhamDev** [comment]
OK, this should be it. Now just squash your commits again.
And here's a tip: Use `git commit --amend` when making changes, so squashing isn't necessary.

**Review by YeldhamDev — APPROVED** [review]

**Comment by Rindbee** [comment]
> OK, this should be it. Now just squash your commits again. And here's a tip: Use `git commit --amend` when making changes, so squashing isn't necessary.

Yeah, But that may have to make a force-push, it doesn't seem to be  easy to see the details.

**Comment by YeldhamDev** [comment]
> Yeah, but that may have to make a force-push, it doesn't seem to be easy to see the details.

With small PRs like this that doesn't really matter.

File: doc/classes/String.xml [hunk 9]
```diff
@@ -181,6 +181,29 @@
...
+				var places = ["GodotDesert", "MountGodot", "GodotRiver", "GodotLake"]
+				# This placeholder is "$_" now.
+				str = str.format(places, "$_") # "The NPC, named Tom, is 24 years old, he comes from MountGodot, and is a member of [organization_name]..."
+
+				# May find the organization info like this.
+				var org_info = [["id", 1 ], ["organization_name", "A"], ["position", 1], ["members", []]]
+				# The placeholder now is "[_]".
+				str = str.format(org_info, "[_]") # "The NPC, named Tom, is 24 years old, he comes from MountGodot, and is a member of A..."
+				[/gdscript]
+				[/codeblocks]
```
Comment (Calinou) [inline doc/classes/String.xml:206]:
```suggestion
				[/codeblocks]
				See [url=$DOCS_URL/tutorials/scripting/gdscript/gdscript_format_string.html#format-method-examples]Format method examples[/url] in the documentation for more information.
```

Comment (Calinou) [inline doc/classes/String.xml:206]: Since there is no C# counterpart, the GDScript-specific codeblock should be removed. Also, singular `[codeblock]` tag should be used instead:

```suggestion
				[codeblock]
				# Like this string, the placeholders may be replaced in steps.
				var str = "The NPC, named {npc_name}, is {npc_age} years old, he comes from $1, and is a member of [organization_name]..."
				# You may already get the npc info like this.
				var npc_info = {"npc_name": "Tom", "npc_age": 24, "organization_id":021, "npc_equipments": {"head": 001, "left_hand": 012}}
				# As the placeholder is the default one, so we can just format `str` like this.
				str = str.format(npc_info) # "The NPC, named Tom, is 24 years old, he comes from $1, and is a member of [organization_name]..."

				# The list of places may look like this.
				var places = ["GodotDesert", "MountGodot", "GodotRiver", "GodotLake"]
				# This placeholder is "$_" now.
				str = str.format(places, "$_") # "The NPC, named Tom, is 24 years old, he comes from MountGodot, and is a member of [organization_name]..."

				# May find the organization info like this.
				var org_info = [["id", 1 ], ["organization_name", "A"], ["position", 1], ["members", []]]
				# The placeholder now is "[_]".
				str = str.format(org_info, "[_]") # "The NPC, named Tom, is 24 years old, he comes from MountGodot, and is a member of A..."
				[/codeblock]
```

Otherwise, no code block will appear at all if the user chose to display only C# code blocks. This can be confusing.

**Review by Calinou — COMMENTED** [review]
On second thought, the code sample actually seems hard to follow for me. I find the comments quite confusing as well, especially since the dictionaries/arrays contain unrelated information. Therefore, I think it would be better to showcase a *minimal* example, even if it's less comprenhensive. Something like:

```gdscript
var string = "The NPC {name} lives in {location}.".format({"name": "Tom", "location": "the village"})
print(string)  # The NPC Tom lives in the village.
```

I wouldn't bother showcasing custom placeholders within the class reference, as it's a very situational use case. 99% of the time, standard named placeholders are all you need, including for internationalization :slightly_smiling_face: 

Also, `str` is not a valid variable name (it's a reserved built-in function), so it should be renamed to `string`.

**Comment by Rindbee** [comment]
Well, indeed, more examples can refer to the tutorial.

File: doc/classes/String.xml [hunk 10]
```diff
@@ -181,6 +181,14 @@
 			<argument index="1" name="placeholder" type="String" default="&quot;{_}&quot;" />
 			<description>
 				Formats the string by replacing all occurrences of [code]placeholder[/code] with [code]values[/code].
+				[b]Note:[/b] There must be one and only one [code]_[/code] in the [code]placeholder[/code].
```
Comment (timothyqiu) [inline doc/classes/String.xml:184]: Multiple underscores are allowed. Each will be replaced by the element key. Although I don't think there's any valid use case for it. The following snippet works:

```gdscript
print("{0000} {1111}".format(["Hello", "World"], "{____}"))  # Hello World
print("{ididid}".format({id=42}, "{___}"))  # 42
```

File: doc/classes/String.xml [hunk 11]
```diff
@@ -181,6 +181,14 @@
...
+				[b]Note:[/b] The [code]values[/code] can be a dictionary, an one-dimensional array, or a two-dimensional array. When it's a two-dimensional array, the length of the inner arrays must be [code]2[/code], or an error will be generated.
```
Comment (timothyqiu) [inline doc/classes/String.xml:185]: `Array` elements are not required to use the same type. So it does not have to be a strictly two-dimensional array.

```gdscript
# Prints: User 42 is Godot.
print("User {0} is {name}.".format([42, ["name", "Godot"]]))
```

File: doc/classes/String.xml [hunk 12]
```diff
@@ -181,6 +181,14 @@
...
+				[b]Note:[/b] The function will iterate over the [code]values[/code]. For a dictionary, first, it will replace the [code]_[/code] with the key; Then use the result to find the first matched word; At last, use the value to replace the result (see [method replace]). For an one-dimensional array, like the dictionary, but the index number acts as the key, the element of the array acts as the value. For a two-dimensional array, the first element of the inner array acts as the key, the second one acts as the value.
```
Comment (timothyqiu) [inline doc/classes/String.xml:186]: > find the first matched word

It uses `replace()` so the function will replace all occurrences of that result.

You can also omit the indices when using an array:

```gdscript
# Prints: User 42 is Godot.
print("User {} is {}.".format([42, "Godot"], "{}"))
```

**Review by timothyqiu — COMMENTED** [review]
Found some missing behaviors in the description.

I submitted my try today (#62439) and didn't know this PR exists. Sorry, I should do a search beforehand. I'll close mine after this PR is updated :)

File: doc/classes/String.xml (hunk 11 above)
Comment (Rindbee) [reply doc/classes/String.xml:185]: It seems quite complicated.
The elements in the outer array can be a mixture of a single `Variant` and a two-element array.
The elements in the inner array are also not specified must be of type `String`.

File: doc/classes/String.xml (hunk 12 above)
Comment (Rindbee) [reply doc/classes/String.xml:186]: Hmm, but this way it's replaced in order.

File: doc/classes/String.xml (hunk 10 above)
Comment (Rindbee) [reply doc/classes/String.xml:184]: It does work. I've written as the example given in the code comments before.

**Comment by Rindbee** [comment]
Perhaps, I should state that this is the recommended usage.
